### DISCUSSION TITLE
recover panics in fyne

### DISCUSSION THREAD

**Discussion by mpkondrashin** [fyne-io/fyne#4519] [Q&A] [answered]
I want my application log all panics (and/or actually recover), including ones generated within fyne operation, i.e. addressing nil pointer when pushing a button.

Can I put my recover() functions in all fyne goroutines?

**Comment by Jacalz** [comment, marked as answer]
We try rigorously to have our code not crash. If it does, it is a bug and something we should fix. Putting `recover()` everywhere is not the correct solution. Open a bug report using the `bug report` issue template if our code is crashing (one for each crash if there are multiple different panics).

If you own code is crashing, rewrite it to not crash. You should write code to avoid panicing and doing illegal operations.  It is also worth noting that recovering from a panic is a slow operation as it has to rewind the stackframes etc.

**Reply by andydotxyz** [reply]
One more thought on this. If tapping a button causes a nil pointer dereference then the issue is almost certainly that you are operating on nil in your code.

**Reply by mpkondrashin** [reply]
It is not the point, which code crushes, is it my code or fyne itself.

For example, on all of my Windows test boxes, fyne programs crush due to absence of openGL. (I am not alone here. For example: https://github.com/fyne-io/fyne/issues/4145).  I have found solution - put opengl32.dll (software implementation of OpenGL) aside to the executable. On the other hand, I do not want to put it by default - writing files can by itself be a problem (who knows what permissions my process has!), so I would extract embedded file only and only if application crushes. In this case, it will minimize probability of any troubles with my customers.

**Reply by Jacalz** [reply]
If you use `recover()`, you need to do it when calling the function that crashes so I think it is out of the question in this case (not having OpenGl happens deep in our internals). Your best bet is probably to have an external script that runs your application, checks the exit status and then restarts the application with the dll in place if it crashed.

Another potential option:

You might be able to intercept the OS signal using https://pkg.go.dev/os/signal and have a goroutine that waits on that crash signal. If you crash, you might be able to put the DLL in place and launch a new process for your exe. However, we are getting really deep into systems level programming here so it will be complicated, if it works. I know something similar works in C++ but I can't promise that it works in Go.

**Reply by andydotxyz** [reply]
On the topic of OpenGL panic it is the only panic that Fyne will cause. It is a result of a complete inability to launch the runtime environment and so there is nothing we can do to present information to the user.

Any other panics are bugs in Fyne like Jacob notes, or an issue in developer code. And yes it does make a difference where it comes from - neither Fyne nor any library should be adding code to recover from issues in developer code.

**Reply by Jacalz** [reply]
I think we panic if ShowAndRun() isn't called from the main goroutine as well but that's a developer error if it happens and should basically never occur.

**Reply by andydotxyz** [reply]
You're absolutely right. Another irrecoverable setup issue.

**Reply by mpkondrashin** [reply]
Thank you very much for all of these ideas!
